ray-curator/ray_curator/stages/deduplication/fuzzy/fuzzy_deduplication.py

```diff
@@ -0,0 +1,295 @@
+import time
+from typing import Any, Literal
+
+from loguru import logger
+
+from ray_curator.backends.experimental.ray_actor_pool import RayActorPoolExecutor
+from ray_curator.pipeline import Pipeline
+from ray_curator.stages.deduplication.fuzzy.buckets_to_edges import BucketsToEdgesStage
+from ray_curator.stages.deduplication.fuzzy.connected_components import ConnectedComponentsStage
+from ray_curator.stages.deduplication.fuzzy.identify_duplicates import IdentifyDuplicatesStage
+from ray_curator.stages.deduplication.fuzzy.lsh.stage import LSHStage
+from ray_curator.stages.deduplication.fuzzy.minhash import MinHashStage
+from ray_curator.stages.deduplication.id_generator import create_id_generator_actor, get_id_generator_actor
+from ray_curator.stages.file_partitioning import FilePartitioningStage
+from ray_curator.tasks import FileGroupTask
+from ray_curator.utils.file_utils import get_fs
+
+
+class FuzzyDeduplicationWorkflow:
+    """
+    A pipeline that performs fuzzy deduplication of a dataset.
+    It consists of the following stages:
+    - FilePartitioningStage
+        Groups input files into smaller groups that can be processed in parallel.
+    - MinHashStage
+        Computes minhashes for the input dataset.
+    - LSHStage
+        Performs Locality Sensitive Hashing on the minhashes.
+        This is a shuffle stage that involves moving data between workers.
+    - BucketsToEdgesStage
+        This stage converts the resulting LSH mapping of bucket ID to document ID into a graph of edges.
+    - ConnectedComponentsStage
+        Performs weaklyconnected components clustering on the graph represented by the edgelist.
+    - GenerateRemovalIDs
+        Generates a list of document ids to remove based on the connected components clusters/components.
+    - Removal (Optional)
+        Currently not implemented.
+    """
+
+    def __init__(  # noqa: PLR0913
+        self,
+        # I/O config
+        input_path: str | list[str],
+        cache_path: str,
+        output_path: str | None = None,
+        input_filetype: Literal["jsonl", "parquet"] = "parquet",
+        input_blocksize: str | int = "1GiB",
+        input_file_extensions: list[str] | None = None,
+        read_kwargs: dict[str, Any] | None = None,
+        cache_kwargs: dict[str, Any] | None = None,
+        write_kwargs: dict[str, Any] | None = None,
```

Comment:
Seems related to this question I had
https://nvidia.slack.com/archives/C098W91J3ND/p1755909117350379

Is there only one cache_kwargs? Do you prefer cache_kwargs or just read/write_kwargs. To mimic, I think we should also have the same for SemDedup

Reply:
I was viewing the pipeline to be as follows:
1. `read_kwargs` kwargs for the input
2. `cache_kwargs` kwargs for all the intermediates (minhashes etc.) written out by dedup. This will suffer from the same issue described on the thread where some stages might want different kwargs than others.
3. `write_kwargs` - This is currently unused because removal is a separate part of the workflow but the goal was to have `output_path` and `write_kwargs` be used for the removal and final write of the resulting dataset.